packages/server/src/core/middleware.ts

```diff
@@ -180,14 +186,34 @@ export function createMiddlewareFactory<TConfig extends AnyRootConfig>() {
   }
 
   function createMiddleware<TNewParams extends ProcedureParams>(
-    fn: MiddlewareFunction<deriveParamsFromConfig<TConfig>, TNewParams>,
-  ): MiddlewareBuilder<deriveParamsFromConfig<TConfig>, TNewParams> {
+    fn: MiddlewareFunction<
+      deriveParamsFromConfig<TConfig, TInputIn>,
+      TNewParams
+    >,
+  ): MiddlewareBuilder<deriveParamsFromConfig<TConfig, TInputIn>, TNewParams> {
     return createMiddlewareInner([fn]);
   }
 
   return createMiddleware;
 }
 
+export const standaloneMiddleware = <
+  TCtx extends {
+    ctx?: object;
+    input?: object;
+  },
+>() => ({
+  create: createMiddlewareFactory<
+    RootConfig<{
+      ctx: TCtx extends { ctx: infer T } ? T : object;
+      meta: object;
+      errorShape: object;
+      transformer: object;
+    }>,
+    TCtx extends { input: infer T } ? T : unknown
+  >(),
+});
```

Comment:
what's the rationale for this being a builder? why can it not just be `standaloneMiddleware((opts) => { ... })`

Reply:
Hey, thanks for the question. Your guess is right: tRPC typings are quite far in the weeds and since this was my first tour around the codebase, using a builder here was the approach that helped get the types working without a lot of type noise. I tried the direct callback approach with types that I assumed were equivalent, but ran into errors that I would need some time to figure out.

I'm still not completely happy with the TSC error messages (i.e. in cases where an error is expected) even for the builder version, but at least I got it to do what I intended it to do.

Reply:
Yes i think @Nick-Lucas tried this before and eventually gave up, but this is definetely something we want to have

Reply:
What's the way forward with this feature? I'll be honest and say I currently don't have a very good idea of why wrapping this in `.create()` results in valid types but using a direct callback doesn't.

I dug up [this old discussion](https://github.com/trpc/trpc/issues/4191) which you were probably referring and definitely agree that this would up the flexibility as far as 3rd party libs etc are concerned.

However I also kind of don't want to integrate a feature where it's unclear why it works one way but doesn't work the other way. I'll probably have to spend more time understanding exactly how all the ProcedureBuilder/MiddlewareFunction/MiddlewareBuilder/ProcedureParams etc. interact together, but right now it's really hard to hold all of it in my head.

Reply:
I think the issue is that with the builder version, `_ctx_out` can be inferred from the provided callback implementation, but in the non-builder version the callback's type has to be annotated, and I can't see a way of doing it so that `_ctx_out` gets a valid type that is inferred from the implementation.